src/include/duckdb/execution/operator/order/physical_order.hpp

```diff
@@ -20,10 +20,12 @@ class OrderGlobalState;
 //! Physically re-orders the input data
 class PhysicalOrder : public PhysicalOperator {
 public:
-	PhysicalOrder(vector<LogicalType> types, vector<BoundOrderByNode> orders, idx_t estimated_cardinality);
+	PhysicalOrder(vector<LogicalType> types, vector<BoundOrderByNode> orders,
+	              vector<unique_ptr<Expression>> projections, idx_t estimated_cardinality);
 
 	//! Input data
 	vector<BoundOrderByNode> orders;
+	vector<unique_ptr<Expression>> projections;
```

Comment:
I'm not quite convinced that it is necessary to add projection columns to the order node itself. Can't we get the same functionality by pushing a projection node before the order in the plan?

Reply:
Let's say we have this query:
```sql
SELECT a
FROM tbl
ORDER BY a, b, c;
```

This plan is generated:
```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             #0            │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          ORDER_BY         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         tbl.a ASC         │
│           b ASC           │
│           c ASC           │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│             c             │
└───────────────────────────┘                   
```

We need columns `a`, `b`, and `c` in the `ORDER_BY` operator, so we can't project any columns out beforehand, so we project them out afterwards. However, this causes the `ORDER_BY` operator to duplicate the `a`, `b`, and `c` columns in both the key and payload columns within the operator.

By pushing the projection into the operator, we will only add `a` to the payload columns, meaning less column duplication.

This optimizer can also be used to push projections into a table scan, which is useful when the table scan is performed by a remote instance. For example, this query:
```sql
EXPLAIN SELECT b, c FROM tbl WHERE a > 42;
```

We get this plan:
```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             b             │
│             c             │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│             c             │
└───────────────────────────┘      
```

While the `SEQ_SCAN` is performed by a remote DuckDB instance, the `PROJECTION` may be performed locally. However, at this point the `a` column travels over the network, even though it is immediately projected out. If we push the projection into the table scan we can prevent this.